### ISSUE TITLE
Wrong behavior when a role is used multiple times

### ISSUE THREAD

**Issue by hoall** [ansible/ansible#22005]
##### ISSUE TYPE

 - Bug Report

##### ANSIBLE VERSION

```
ansible 2.2.1.0
  config file = /home/me/.ansible.cfg
  configured module search path = Default w/o overrides

```

##### CONFIGURATION

I set the **roles_path** to my user account to be able to run ansible-galaxy install -r requirements.yml

##### OS / ENVIRONMENT

N/A

##### SUMMARY

When using a role multiple times in a playbook with other roles in between it breaks the roles order and drops the given variable parameter.

I'm pretty sure that this worked in earlier versions.

##### STEPS TO REPRODUCE

```yaml
---

- hosts: example
  roles:
  - { role: role_a }
  - { role: role_b }
  - { role: role_c, variable1: foo, variable2: 1234 }
  - { role: role_d }
  - { role: role_c, variable1: bar, variable2: 5678 }
  - { role: role_d }
```

##### EXPECTED RESULTS

The role execution should keep the given order. role_d should be called twice, each time with the given variables.

##### ACTUAL RESULTS

Ansible keeps the order until the point it finished role_c. Than for a reason I don't know, it skips role role_d and starts role_c the second time. The second run fails due to the fact, that ansible seems to forget about the given variables.

The output below is my actual output, I just replaced my account name and the role name.

```
fatal: [myhost]: FAILED! => {"failed": true, "msg": "the field 'args' has an invalid value, which appears to include a variable that is undefined. The error was: 'variable1' is undefined\n\nThe error appears to have been in '/home/me/.ansible/roles/role_c/tasks/main.yml': line 20, column 3, but may\nbe elsewhere in the file depending on the exact syntax problem.\n\nThe offending line appears to be:\n\n\n- name: Unpack foo\n  ^ here\n"}
```

**Comment by bcoca** [comment]
have you set allow_duplicates to True?

**Comment by bjolivot** [comment]
I tried same thing in 2.3.
Result was
role_a
role_b
role_c
role_d
role_c

(role_d was played only once)

**Comment by bcoca** [comment]
you still have not answered my question

**Comment by hoall** [comment]
@bcoca Sorry for the delay, I'm living in germany.

No I hadn't. I added it now role_c/meta/main.yml, but's still the same result.

@bjolivot Thanks for testing, but I'm still on 2.2.1.  I expect you mean **role_c** is only run once? that would be another behavior than I experienced.

@bcoca Is **allow_duplicates** the right approach for me? My dupblicates a direct role calls and no role dependencies in a classic way.

Also I'm sure that in an earlier version this worked.

Thank you for you support.
